### PR TITLE
feat: add support for typeddicts

### PR DESCRIPTION
**Pull request by ivanleomk**
This addresses #627 and adds support for typed dicts in Instructor. Users can now use the following syntax

```python
from typing_extensions import TypedDict
from openai import OpenAI
import instructor


class User(TypedDict):
    name: str
    age: int


client = instructor.from_openai(OpenAI())

print(
    client.chat.completions.create(
        model="gpt-3.5-turbo",
        response_model=User,
        messages=[
            {
                "role": "user",
                "content": "Timothy is a man from New York who is turning 32 this year",
            }
        ],
    )
)

"""
name='Timothy' age=32
"""
```

----

| :rocket: | This description was created by [Ellipsis](https://www.ellipsis.dev) for commit e7341fd0d4489e6fc25834bf81071b023f248fd3  | 
|--------|--------|

### Summary:
This PR adds support for TypedDict in Instructor by converting them to Pydantic models and includes comprehensive tests for these features.

**Key points**:
- Added support for `TypedDict` in `Instructor` by converting them to Pydantic models.
- Modified `tests/llm/test_openai/test_patch.py` to include tests for `TypedDict` support.
- Updated `test_typed_dict` to assert the response is an instance of `BaseModel` instead of `UserExtract`.

----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)

### CODE REVIEW

File: tests/llm/test_openai/test_patch.py [hunk 1]
```diff
@@ -12,6 +13,30 @@ class UserExtract(BaseModel):
...
+    client = instructor.patch(client, mode=mode)
+    model = client.chat.completions.create(
+        model=model,
+        response_model=UserExtractTypedDict,
+        max_retries=2,
+        messages=[
+            {"role": "user", "content": "Extract jason is 25 years old"},
+        ],
+    )
+    assert isinstance(model, BaseModel), "Should be instance of a pydantic model"
```
[the remark this answers is not in the document]
Comment (ivanleomk) [reply tests/llm/test_openai/test_patch.py:32]: We can only assert that it's an instance of a BaseModel...

**Comment by ivanleomk** [comment]
Small downside is that we don't get type hints anymore but that is expected since we are using the `create_model` method to take in arbitrary typed dicts

File: tests/test_process_response.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+def test_typed_dict_conversion() -> None:
+    class User(TypedDict):  # type: ignore
+        name: str
+        age: int
+
+    _, user_tool_definition = handle_response_model(User)
+
+    class User(BaseModel):
+        name: str
+        age: int
```
Comment (ADR-007) [inline tests/test_process_response.py:15]: Does it work with Required/NotRequired types? 🤔

File: instructor/process_response.py [hunk 3]
```diff
@@ -193,6 +201,12 @@ def handle_response_model(
         if is_simple_type(response_model):
             response_model = ModelAdapter[response_model]
 
+        if is_typed_dict(response_model):
+            response_model: BaseModel = create_model(
```
Comment (ADR-007) [inline instructor/process_response.py:205]: Maybe it would be easier to use RootSchema?
```
response_model = RootModel[response_model]
```

Comment (ADR-007) [reply instructor/process_response.py:205]: But in this case, the value should taken from `response_model.root`, so we need to have another `if` 🤔

Comment (ADR-007) [reply instructor/process_response.py:205]: Nevermind, it doesn't work

File: tests/test_process_response.py (hunk 2 above)
Comment (ivanleomk) [reply tests/test_process_response.py:15]: Hmm currently with my approach, I assume all keys are required but I see where you're coming from. Will try to see what I can do

Comment (jxnl) [reply tests/test_process_response.py:15]: is optional not supported?

**Review by jxnl — APPROVED** [review]
Please make sure to add this in concepts. typeddict.md or something in concepts

**Pull request merged**